execute/executor.go

```diff
@@ -298,6 +295,61 @@ func (v *createExecutionNodeVisitor) Visit(node plan.Node) error {
 	return nil
 }
 
+func (es *executionState) validate() error {
+	if es.resources.ConcurrencyQuota == 0 {
+		return errors.New(codes.Invalid, "plan must have a non-zero concurrency quota")
+	}
+	return nil
+}
+
+
+func (es *executionState) chooseDefaultResources(ctx context.Context, p *plan.Spec) {
+	// Initialize resources from the execution dependencies and/or properties of the plan.
+	if ! HaveExecutionDependencies(ctx) {
+		return
+	}
+
+	execOptions := GetExecutionDependencies(ctx).ExecutionOptions
+
+	// Update memory quota
+	if es.resources.MemoryBytesQuota == 0 {
+		es.resources.MemoryBytesQuota = execOptions.DefaultMemoryLimit
+	}
+
+	// Update concurrency quota
+	if es.resources.ConcurrencyQuota == 0 {
+		es.resources.ConcurrencyQuota = len(p.Roots)
+
+		// If the query concurrency limit is greater than zero,
+		// we will use the new behavior that sets the concurrency
+		// quota equal to the number of transformations and limits
+		// it to the value specified.
+		if execOptions.ConcurrencyLimit > 0 {
+			concurrencyQuota := 0
+			_ = p.TopDownWalk(func(node plan.Node) error {
+				// Do not include source nodes in the node list as
+				// they do not use the dispatcher.
+				if len(node.Predecessors()) > 0 {
+					addend := 1
+					ppn := node.(*plan.PhysicalPlanNode)
+					if attr, ok := ppn.OutputAttrs[plan.ParallelRunKey]; ok {
+						addend = attr.(plan.ParallelRunAttribute).Factor
+					}
+					concurrencyQuota += addend
+				}
+				return nil
+			})
+
+			if concurrencyQuota > int(execOptions.ConcurrencyLimit) {
+				concurrencyQuota = int(execOptions.ConcurrencyLimit)
+			} else if concurrencyQuota == 0 {
+				concurrencyQuota = 1
```

Comment:
Curious about this branch here. If we had a trivial query whose execution graph just had a `ReadWindowAggregate` node, I guess `concurrencyQuota` could be 0. Do we require it to be positive even if there are no non-sources?

Reply:
I might be mistaken, but my guess is there needs to be at least one goroutine to work the consecutive transport belonging to the source nodes. The source nodes themselves I think just deposit messages to the outgoing dataset and that's as far as the source goroutines take it. There needs to be a dispatcher thread that reads those messages and writes to CSV writer.

Reply:
Anecdotally, during recent refactors when I failed to mark any nodes as roots (my mistake) the concurrency quota was set to zero, which produced an error. It didn't seem like there was any conditional around that check since it failed for from/range/filter which I think would have be rewritten as a single source.

https://github.com/influxdata/flux/blob/master/execute/executor.go#L80-L85

Reply:
Right makes sense.

Given Go's convention of having a useful zero value, I wonder if we should change the meaning of concurrency quota to be the number of additional goroutines after the required one. Or maybe `0` should just mean the default of `1`.

Nothing that needs to change here for this PR, it just seems a little weird.

Reply:
Right makes sense.

Given Go's convention of having a useful zero value, I wonder if we should change the meaning of concurrency quota to be the number of additional goroutines after the required one. Or maybe `0` should just mean the default of `1`.

Nothing that needs to change here for this PR, it just seems a little weird.